gateway/websocket_request_context.go

```diff
@@ -146,99 +142,62 @@ func (wrc *websocketRequestContext) buildProtocolContextFromHTTPRequest(
-	<-completionChan
+	// Create separate channels for establishment and closure observations
+	establishmentObservationsChan := make(chan *protocolobservations.Observations, 1)
+	closureObservationsChan := make(chan *protocolobservations.Observations, 1)
 
-	logger.Info().Msg("🔌 WebSocket connection terminated, broadcasting final connection observations")
+	// Start listening for connection observations in separate goroutines
+	go wrc.listenForEstablishmentObservations(establishmentObservationsChan)
+	go wrc.listenForClosureObservations(closureObservationsChan)
 
-	return nil
-}
-
-// startWebSocketBridge creates a websocket bridge and starts it.
-// It also starts a goroutine to listen for message processing notifications from the bridge.
-// This method returns a completion channel that signals when the bridge shuts down.
-func (wrc *websocketRequestContext) startWebSocketBridge(
-	httpRequest *http.Request,
-	httpResponseWriter http.ResponseWriter,
-) (<-chan struct{}, error) {
-	logger := wrc.logger.With("method", "startWebSocketBridge")
-
-	// Get the websocket-specific URL from the selected endpoint.
-	websocketEndpointURL, err := wrc.protocolCtx.GetWebsocketEndpointURL()
-	if err != nil {
-		// Wrap the endpoint URL error with our specific error type
-		endpointErr := fmt.Errorf("%w: selected endpoint does not support websocket RPC type: %s", errWebsocketConnectionFailed, err.Error())
-		wrc.updateGatewayObservations(endpointErr)
-		logger.Error().Err(err).Msg("❌ Selected endpoint does not support websocket RPC type")
-		return nil, endpointErr
-	}
-	wrc.logger = wrc.logger.With("websocket_url", websocketEndpointURL)
-
-	// Get the headers for the websocket connection that will be sent to the endpoint.
-	endpointConnectionHeaders, err := wrc.protocolCtx.GetWebsocketConnectionHeaders()
-	if err != nil {
-		// Wrap the connection headers error with our specific error type
-		headersErr := fmt.Errorf("%w: failed to get websocket connection headers: %s", errWebsocketConnectionFailed, err.Error())
-		wrc.updateGatewayObservations(headersErr)
-		logger.Error().Err(err).Msg("❌ Failed to get websocket connection headers")
-		return nil, headersErr
-	}
-
-	// Start the websocket bridge and get a completion channel.
-	// The bridge uses the websocket request context as the message processor to
-	// perform both protocol-level and QoS-level message processing.
-	// Pass the shared WebSocket context so both bridge and gateway use the same lifecycle.
-	completionChan, err := websockets.StartBridge(
+	// Delegate to the protocol context to start the WebSocket bridge
+	// The protocol context handles all protocol-specific setup
+	// We pass ourselves (wrc) as the messageProcessor to handle message processing
+	err := wrc.protocolCtx.StartWebSocketBridge(
 		wrc.context,
-		wrc.logger,
 		httpRequest,
 		httpResponseWriter,
-		websocketEndpointURL,
-		endpointConnectionHeaders,
-		wrc,
+		wrc, // Pass the websocketRequestContext as the message processor
```

Comment:
Message processor is quite confusing here: please use a go-idiomatic name, i.e. focus on what the method/struct actually does.

Meta point: this needs a few refactors to establish the gateway package's context as the coordinator of operations: e.g. 
- receive messages from the protocol context over a channel -> pass to QoS -> use QoS context to write user's response, etc.
- receive observations from the protocol context over a channel -> hand over to publisher(s) like data, metrics, etc.

Reply:
A TODO works, thank you!